### ISSUE TITLE
Defining time steps for a GCM using a standard 360 days calendar

### ISSUE THREAD

**Issue by farinfa** [ec-jrc/lisflood-lisvap#28]
Hi,

I'm having some troubles running LISVAP over a set of climate variables derived from the HadGem2-ES projections. This GCM is structured to have a standardized 360 days per year calendar. The full time series spans the period 01/12/1949 - 30/11/2099 (don't ask me why these specific dates...) for a total of 54000 time steps (360x150). 
Now, if I set LISVAP with the start and end dates of the HadGem dataset, I would need to define 55109 time steps, but this would be inconsistent with the input files (54000 time steps). If I respect the 54000 time steps, on the other hand, I would have to adjust somehow the start and end dates.
How should I deal with this issue? 

![image](https://user-images.githubusercontent.com/62378247/77842765-7fb7b380-7196-11ea-9fea-8f2e858a18ea.png)
![image](https://user-images.githubusercontent.com/62378247/77842776-af66bb80-7196-11ea-90d3-b5f81a41de48.png)

Thanks a lot for your help!

**Comment by domeniconappo** [comment]
@farinfa 

Calendar management is weak in lisvap, indeed. 
As I remember, there was some effort on Lisflood to tackle different calendars but it wasn't done in lisvap (unfortunately, originally those two software, although very similar in code in many parts, weren't developed to avoid code duplication, and this is a typical effect of that).

Now I need to replicate the work done on Lisflood into lisvap. It won't be quick because I need to be careful to avoid regression or strange effects that can be hard to spot and debug...

Two questions, while I start to work on it:

1. Does your inputs define "calendar"  attribute for their time variable? Netcdf time variables accept an attribute "calendar"...check if it's defined.Yours should be `360_day`.

https://www.unidata.ucar.edu/software/netcdf-java/current/CDM/CalendarDateTime.html

2. Have you tried to define StepStart and StepEnd as integers? In Lisvap/Lisflood you can use integer timesteps.

```xml
<textvar name="StepStart" value="1" />
<textvar name="StepEnd" value="54000" />
```
If this seems to work,  calendar in output files is fixed to be `gregorian` so to have proper dates in  output you should change time.calendar in netcdf results...

**Comment by farinfa** [comment]
Hi @domeniconappo,

Thanks a lot for your help. It would be indeed a very nice development of the model, but for the moment being, I think we can edit the output files manually.

Coming to your questions:
1. The input files do have the `360_day` calendar defined (see the image)
![image](https://user-images.githubusercontent.com/62378247/77891379-25d7ec00-7271-11ea-9a3e-fa84a780a3b1.png)

2. No, I have not tried to use integers, but I'll try this way.

Thanks
F

**Comment by farinfa** [comment]
Ciao @domeniconappo,

On the same line, I would like to bring to your attention an additional issue. While with the `360_day` the model generates an output, somehow filling the days where the inputs are not available, with a `365_day` calendar it exits the process with an error.
![image](https://user-images.githubusercontent.com/62378247/78111008-47b1aa00-73fc-11ea-88b4-9f9cb4d816c3.png)
![image](https://user-images.githubusercontent.com/62378247/78111030-55ffc600-73fc-11ea-8af4-1454d4f5ac3b.png)
Thanks again for your support

Fabio

**Comment by emiliano-gelati** [comment]
Hi @domeniconappo ,
(cc: @farinfa , @AdDeRoo , @bissebe )

If you have not started working on this, I can help perhaps.

The issue seems to be the same as [the one fixed for LISFLOOD](https://github.com/ec-jrc/lisflood-code/commit/e5eb9f03da221bbe5f1da76c9d9f1c83afcb38f9).

On the one hand, the date parser in the "calendar" function (https://github.com/ec-jrc/lisflood-lisvap/blob/master/src/lisvap/utils/tools.py#L83) does not take the [CF calendar type](http://cfconventions.org/Data/cf-conventions/cf-conventions-1.8/cf-conventions.html#calendar) into account. To convert date strings to integers, [cftime.date2num](https://unidata.github.io/cftime/api.html) could be used instead.

On the other hand, [the hard-coded calendar type in the NetCDF writing function](https://github.com/ec-jrc/lisflood-lisvap/blob/master/src/lisvap/utils/writers.py#L79) should be replaced with the correct calendar type string, which should be read as the "calendar" attribute of the NetCDF "time" variable of some forcing file or provided in the settings XML (with the due consistency checks).

**Comment by domeniconappo** [comment]
Hi @emiliano-gelati 
thanks your inputs! Indeed, I took exactly the same road, and changed hard-coded calendar and so on.

BTW, one source of confusion that is somehow (not strictily) related to this issue, is the timestepflag='closest' as default in readnetcdf.
https://github.com/ec-jrc/lisflood-lisvap/blob/master/src/lisvap/utils/readers.py#L162

This is not the case in lisflood (where timestepflag is 'exact' by default).
By changing that default value from `closest` to `exact` it makes one test fail. But it seems related to the dataset I used for that test (where units are not homogeneous between inputs).

 I will let you review my changes if you have some time to spend on it.

**Comment by emiliano-gelati** [comment]
Hi @domeniconappo ,

I am not sure about the timestampflag='closest' default, but my guess would be that it was set to read forcing variables available at daily resolution in a 6-hourly run.

I will be glad to review your changes or help.

**Comment by farinfa** [comment]
Just for the record, meanwhile you are working in a more appropriate solution to this problem, I have found a workaround to get the model to work with the [365_day calendar](https://github.com/ec-jrc/lisflood-lisvap/issues/28#issuecomment-607086705) dataset. 

I have changed the calendar manually to a `proleptic_gregorian` and made `CDO` re-date all the time steps. 
`cdo -settaxis,1951-01-01,00:00:00,1day -setcalendar,proleptic_gregorian infile_365.nc outfile_gregorian.nc`
In this way, for all the leap years the timeseries "shift to a day before", which is not optimal, but at the end is only a 37 days shift over 150 years. 
Of course, I'll re run LISVAP once a fix will be made available.
Thank you very much for your help

F

**Comment by domeniconappo** [comment]
HI @farinfa 
thanks for sharing. 

Yes, manipulating input dataset with a tool like cdo is a workaround. Of course we want to avoid that burden for the user. Probably the new version will be available next week!

**Comment by gnrgomes** [comment]
Hi @farinfa 
Is this issue also solved by issue "**#43: 365_day calendar in Lisvap**"?
Could you please test it using the new version: lisflood-lisvap==1.0.4 ?
Please, give me feedback so we can close also this issue if everything is ok.

**Comment by farinfa** [comment]
Hi @gnrgomes,

I have downloaded the new lisvap version: I'll test both the issues and confirm.
Thank you
